customers.md

```diff
@@ -1,41 +1,41 @@
 # Customers
 
-Customers are stored in Craft Commerce for the purpose of having a single place to relate orders to.
+With Craft Commerce, a _Customer_ is a model representing a person who placed an order.
 
-A customer will always be created during checkout if they are a guest, or if the user is logged in, the customer related to the logged in user will be used.
+That person may have placed an order as a guest or logged in with an account via a related [user account](https://docs.craftcms.com/v3/users.html).
 
-Logged-in users can only be a single customer in the system. A person may checkout as a guest customer with the same email multiple times, and a new guest customer will be created each time. After an ordered is completed customers are consolidated so there is only ever one customer per email address.
+Commerce will consolidate customers by email address. A person may checkout as a guest customer with the same email multiple times, and a new guest customer will be created each time. After an order is completed, customers are consolidated so there is only ever one customer per email address.
 
-After a user is registered and verified, whenever they log in, all guest orders that used the email address belonging to that user are transferred to the logged in user. This means, even if a registered user makes an anonymous/guest order, the next time they log in, the order will appear in their order history, as long as they used the same email address.
+Once a user is registered and verified, orders on their email address will be consolidated whenever they log in. Any guest orders with the user’s email address are transferred to the user. As a result, even future anonymous/guest orders with that email address will appear in the user’s order history.
```

Comment:
This has changed in Commerce 3 and we haven‘t updated this.

How it works is the following:

All users in the system have an associated customer record. Then when a logged in user is building a cart on the front end, the customerId of the order/cart is always the customerId belonging to that user.

If a guest customer creates an order, a new guest customer is created for that order. When they complete that order, if there was a registered user with the same email address as the order, the order is moved to belong to the customerId of the registered user.
If there was not a registered user with the same email, but their was a previous guest customer with the same email address, the order is moved to belong to that single guest customer.

Everything now happens on order completion.

The above system means that guest customers that had their orders move to another customer, are now orphaned, and we garbage collect them (delete them) often.

let me know if you need more clarification @mattstein

Reply:
Thanks Luke! This makes sense to me except for one part:

> All registered users in the system have an associated customer record.

Are you saying that a) every User now inherently has an associated Customer by default, or that b) every visitor that’s created a store account has a Customer record with an associated Craft user (as it was before)?

It seems like you’ve mostly described a change to the Customer consolidation strategy that’s now orphan-and-garbage-collect, and that the moment of tidying up is now on order completion and not login. Is that accurate?

Reply:
In Commerce 3, anytime a Craft user element is saved, we make sure a related customer record exists for that user, and create one for that user if there wasn't one.

In the Commerce 3 migration (developers moving from commerce 2) we generated a customer record for every Craft user element that didnt have one.


> It seems like you’ve mostly described a change to the Customer consolidation strategy that’s now orphan-and-garbage-collect, and that the moment of tidying up is now on order completion and not login. Is that accurate?

Exactly.

Reply:
Cool, then it sounds like it’s _a)_, where a tight User<->Customer relationship is a careful design choice in v3. And the Craft user has a Customer record even if they never created an order.

This is all making sense, thanks for clarifying!